raiden/waiting.py

```diff
@@ -369,38 +372,51 @@ def wait_for_healthy(
     wait_for_network_state(raiden, node_address, NODE_NETWORK_REACHABLE, retry_timeout)
 
 
-def wait_for_transfer_success(
+class TransferWaitResult(Enum):
+    SECRET_REGISTERED_ONCHAIN = "secret registered onchain"
+    UNLOCKED = "unlocked"
+    UNLOCK_FAILED = "unlock_failed"
+
+
+def wait_for_transfer_result(
     raiden: "RaidenService",
     payment_identifier: PaymentID,
     amount: PaymentAmount,
     retry_timeout: float,
-) -> None:  # pragma: no unittest
-    """Wait until a transfer with a specific identifier and amount
-    is seen in the WAL.
+    secret: Secret,
+) -> TransferWaitResult:  # pragma: no unittest
+    """Wait for the result of a transfer with the specified identifier or
+    secret. Possible results are onchain secret registration, successful
+    unlock and failed unlock. For a successful unlock, the amount is also
+    checked.
 
     Note:
         This does not time out, use gevent.Timeout.
     """
     log_details = {"payment_identifier": payment_identifier, "amount": amount}
-    found = False
-    while not found:
+    result = None
+    while result is None:
         assert raiden, TRANSPORT_ERROR_MSG
         assert raiden.wal, TRANSPORT_ERROR_MSG
         assert raiden.transport, TRANSPORT_ERROR_MSG
 
         state_events = raiden.wal.storage.get_events()
         for event in state_events:
-            found = (
-                isinstance(event, EventPaymentReceivedSuccess)
-                and event.identifier == payment_identifier
-                and event.amount == amount
-            )
-            if found:
-                break
-
-        log.debug("wait_for_transfer_success", **log_details)
+            if isinstance(event, EventPaymentReceivedSuccess):
+                if event.identifier == payment_identifier and event.amount == amount:
+                    result = TransferWaitResult.UNLOCKED
+            elif isinstance(event, EventUnlockClaimFailed):
+                if event.identifier == payment_identifier:
+                    result = TransferWaitResult.UNLOCK_FAILED
+            elif isinstance(event, ContractSendSecretReveal):
+                if event.secret == secret:
```

Comment:
This is not really useful without knowing the lock expiration. The secret can be registered either before or after the lock expires, in the first case it is a successful transfer, in the later case it is a failure.

Reply:
In the tests where we use this we want to assert a successful transfer that stays offchain, so it's still correct for this case (but should maybe be renamed to something other than transfer result.)  Since this also introduces the problem of having only the secret to identify relevant events, is there a better alternative (listening to some other event?)

Reply:
This should test for `ContractReceiveSecretReveal` state change instead of `ContractSendSecretReveal` event. The state change has the block at which the secret was registered, which should be compared against the `lock.expiration`.

Reply:
Just to add, a `ContractReceiveSecretReveal` *after* the `lock.expiration` is of no interest (This case is covered by `EventUnlockClaimFailed`).